fix(migrations): handle reused templates in control flow migration

aparzi opened the pull request:
The control flow migration was incorrectly removing `ng-template` elements in scenarios where they were referenced by multiple `*ngIf` directives' `else` clauses and also used independently via `ngTemplateOutlet`.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/contributing-docs/commit-message-guidelines.md
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [X] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.dev application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Template:
```
<div *ngIf="1 == 1; else elseTemplate">
  <h1>TEST</h1>
</div>
<div *ngIf="1 == 1; else elseTemplate">
  <h1>TEST</h1>
</div>

<ng-container [ngTemplateOutlet]="elseTemplate"></ng-container>
<ng-template #elseTemplate>
  <h1>Test</h1>
  <div>Test</div>
</ng-template>
```

After:

```
@if (1==1) {
  <div>
    <h1>TEST</h1>
  </div>
} @else {
  <h1>Test</h1>
  <div>Test</div>
}

@if (1==1) {
  <div>
    <h1>TEST</h1>
  </div>
} @else {
  <h1>Test</h1>
  <div>Test</div>
}
<ng-container [ngTemplateOutlet]="elseTemplate"></ng-container>
```


## What is the new behavior?
This change updates the template processing logic to correctly detect all usages of a template, including property-bound `[ngTemplateOutlet]`. The migration will now preserve the `ng-template` if it's still referenced after the `*ngIf` directives have been migrated.

Now:

```
@if (1 == 1) {
<div><h1>TEST</h1></div>
} @else {
<h1>Test</h1>
<div>Test</div>
} @if (1 == 1) {
<div><h1>TEST</h1></div>
} @else {
<h1>Test</h1>
<div>Test</div>
} <ng-container [ngTemplateOutlet]="elseTemplate"></ng-container>
<ng-template #elseTemplate>
  <h1>Test</h1>
  <div>Test</div>
</ng-template>

```


## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

JeanMeche commented on packages/core/schematics/migrations/control-flow-migration/util.ts:
```diff
@@ -468,9 +468,12 @@ export function processNgTemplates(
-      const replaceRegex = new RegExp(getPlaceholder(name.slice(1)), 'g');
-      const forRegex = new RegExp(getPlaceholder(name.slice(1), PlaceholderKind.Alternate), 'g');
+    for (const [nameWithHash, t] of templates) {
+      const replaceRegex = new RegExp(getPlaceholder(nameWithHash.slice(1)), 'g');
```
nit:
```suggestion
      const name = nameWithHash.slice(1);
      const replaceRegex = new RegExp(getPlaceholder(name), 'g');
```
Since we're doing this operation 3 times

aparzi committed:
fix(migrations): handle reused templates in control flow migration

The control flow migration was incorrectly removing `ng-template` elements in scenarios where they were referenced by multiple `*ngIf` directives' `else` clauses and also used independently via `ngTemplateOutlet`.

thePunderWoman commented:
This PR was merged into the repository. The changes were merged into the following branches:

- main: b96afb4bfc295583c8d0c60156dd8757c7766740
- 20.3.x: 3ebaeccb466119ee43eeaa486f5e132c85e9caa2
